File: C++/프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer, target;
vector<int> numbers;

void dfs(int res, int index) {
    if(index == numbers.size()) {
        answer += (target == res) ? 1 : 0;
        return;
    }
    dfs(res + numbers[index], index + 1);
    dfs(res - numbers[index], index + 1);
}

int solution(vector<int> numbers, int target) {
    ::target = target;
    ::numbers = numbers;
    dfs(0, 0);
    return answer;
}

File: Python3/프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    def dfs(res, index):
        if index == len(numbers):
            nonlocal answer
            answer += int(res == target)
            return
        dfs(res + numbers[index], index + 1)
        dfs(res - numbers[index], index + 1)
    dfs(0, 0)
    return answer